test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -10556,6 +10556,393 @@ BOOST_AUTO_TEST_CASE(snark)
 	BOOST_CHECK(callContractFunction("verifyTx()") == encodeArgs(true));
 }
 
+BOOST_AUTO_TEST_CASE(ufixed_retrieval)
+{
+	char const* sourceCode = R"(
+	contract test {
+		ufixed a = 3.14;
+
+		function f() public returns (ufixed) {
+			return a;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(createFixed(314, 100, 19)));
+}
+
+BOOST_AUTO_TEST_CASE(fixed_retrieval)
+{
+	char const* sourceCode = R"(
+	contract test {
+		fixed a = -3.14;
+
+		function f() public returns (fixed) {
+			return a;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	bytes const &result = callContractFunction("f()");
+	BOOST_CHECK(result == encodeArgs(createFixed(-314, 100, 19)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_echo)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f(ufixed a) public returns (ufixed) {
+			return a;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(ufixed128x19)", [](u256 const &a) -> u256
+	{ return a; }, 0, 100);
+}
+
+
+BOOST_AUTO_TEST_CASE(fixed_echo)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f(fixed a) public returns (fixed) {
+			return a;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed128x19)", [](u256 const &a) -> u256
+	{ return a; }, -50, 50);
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_constant_addition)
+{
+	char const* sourceCode = R"(
+	contract test {
+		function f() public returns (ufixed) {
+			return 1.25 + 2.5;
+		}
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(createFixed(375, 100, 19)));
```

Comment:
There is a macro called `ABI_CHECK` which produces better formatted output in case of an error, please use that.